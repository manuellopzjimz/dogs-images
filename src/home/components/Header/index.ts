import styled from 'styled-components';

export const Header = styled.header`
    position: fixed;
    top: 0px;
    width: 100%;
    height: 64px;
    border-bottom: 3px solid black;
    background-color: ${props => props.theme.backgroundColor};
    display: flex; 
    justify-content: space-around;
    align-items: center;
    padding: 10px;
    box-shadow: 0px 0px 8px rgba(0,0,0,.6);
`;