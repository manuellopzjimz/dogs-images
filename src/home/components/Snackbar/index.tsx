import { homeActions } from "@Home/store";
import { selectErrorFlag, selectErrorMessage } from "@Home/store/selectors";
import React, { useEffect } from "react";
import { useTranslation } from "react-i18next";
import { useDispatch, useSelector } from "react-redux";
import { Container } from "./style";

export const Snackbar = () => {
  const isError = useSelector(selectErrorFlag);
  const errorMessage = useSelector(selectErrorMessage);
  const { t } = useTranslation("home");
  const dispatch = useDispatch();

  useEffect(() => {
    let timeout = setTimeout(() => {
      dispatch(homeActions.resetError());
    }, 5000);

    return () => clearTimeout(timeout);
  }, [isError, errorMessage]);

  return isError ? (
    <Container isVisible={isError}>{t(errorMessage)}</Container>
  ) : null;
};
