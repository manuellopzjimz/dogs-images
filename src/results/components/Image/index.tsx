import React, { useEffect, useRef, useState } from "react";

import { StyledImage, Loader } from "./style";

type Props = {
  url: string;
  alternativeText: string;
};

export const Image = (props: Props) => {
  const [canDisplayImage, setFlag] = useState<boolean>(false);
  const image = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const callback: IntersectionObserverCallback = (entries) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          setFlag(true);
        }
      });
    };

    const observer = new IntersectionObserver(callback, {
      threshold: 1.0,
    });
    if (image.current !== null) {
      observer.observe(image.current);
    }

    return () => observer.disconnect();
  }, [image.current]);

  return canDisplayImage ? (
    <StyledImage
      src={props.url}
      alt={props.alternativeText}
      width={200}
      height={200}
    />
  ) : (
    <Loader ref={image} />
  );
};
