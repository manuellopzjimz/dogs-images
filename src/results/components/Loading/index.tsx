import React from "react";
import Lottie from "react-lottie";

import Animation from "./loading-animation.json";

export const Loading = () => {
  return (
    <Lottie
      height={400}
      width={400}
      options={{
        loop: true,
        autoplay: true,
        animationData: Animation,
        rendererSettings: {
          preserveAspectRatio: "xMidYMid slice",
        },
      }}
    />
  );
};
