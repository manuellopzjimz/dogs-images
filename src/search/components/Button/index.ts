import React from 'react';
import styled from 'styled-components';

type AvailabilityProp = {
    isDisable?: boolean;
};

export const Button = styled.input.attrs({type: 'submit'})`
    width: 300px;
    height: 64px;
    background-color: ${(props: React.HTMLProps<HTMLInputElement> & AvailabilityProp) => props.isDisable ? '#aaa': 'cyan'};
    cursor: ${(props: React.HTMLProps<HTMLInputElement> & AvailabilityProp) => props.isDisable ? 'not-allowed' : 'pointer'};
    border: 1px solid #bbb;
    border-radius: 50px;
    font-weight: bold;
    font-size: 18px;
`;