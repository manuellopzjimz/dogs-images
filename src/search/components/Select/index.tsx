import React, { useState } from "react";

import { Option } from "@Search/types/option";
import { Container, Label, ListItem, Item } from "./style";

export type Props = {
  placeholder: string;
  options: Option[];
  onSelect: Function;
  onBlur: Function;
  value?: string;
};

export const Select = (props: Props) => {
  const [isVisible, setVisibility] = useState<boolean>(false);
  const [selectedValue, selectValue] = useState<string | undefined>(
    props.value
  );

  const selectOption = (option: Option) => {
    props.onSelect(option.value);
    props.onBlur();
    selectValue(option.value);
    setVisibility(false);
  };

  return (
    <Container>
      <Label
        isVisible={isVisible}
        onClick={() =>
          setVisibility((previousVisibility) => !previousVisibility)
        }
      >
        {props.placeholder}
      </Label>
      {isVisible ? (
        <ListItem>
          {props.options.map((option: Option) => (
            <Item
              isSelected={option.value === selectedValue}
              onClick={() => selectOption(option)}
              key={option.value}
            >
              {option.name}
            </Item>
          ))}
        </ListItem>
      ) : null}
    </Container>
  );
};
