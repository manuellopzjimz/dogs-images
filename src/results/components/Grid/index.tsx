import { useTranslation } from "react-i18next";
import React from "react";
import { useSelector } from "react-redux";

import {
  selectLoadingFlag,
  selectPhotos,
} from "@DisplayResults/store/selectors";
import { selectSelectedBreed } from "@Search/store/selectors";
import { Image } from "../Image";
import { Loading } from "../Loading";
import { Container, Label } from "./style";

export const Grid = () => {
  const { t } = useTranslation("results");
  const isLoading = useSelector(selectLoadingFlag);
  const photos = useSelector(selectPhotos);
  const selectedBreed = useSelector(selectSelectedBreed);

  return isLoading ? (
    <Loading />
  ) : (
    <>
      <Label>{t("numberOfResults", { count: photos.length })}</Label>
      <Container>
        {photos.map((photo: string, index: number) => (
          <Image
            key={index}
            url={photo}
            alternativeText={t("alternativeTextPhoto", {
              number: index,
              breed: selectedBreed,
            })}
          />
        ))}
      </Container>
    </>
  );
};
