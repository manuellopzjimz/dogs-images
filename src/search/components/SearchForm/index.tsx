import React, { useEffect } from "react";
import { useForm, Controller, FieldError } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { yupResolver } from "@hookform/resolvers/yup";
import { useDispatch, useSelector } from "react-redux";

import { Form } from "@Search/types/form";
import { StyledForm, SelectContainer } from "./style";
import { SearchValidator } from "@Search/validators/search";
import { Button } from "../Button";
import { Select } from "../Select";
import {
  areBreedsAlreadyRequested as areBreedsAlreadyRequestedSelector,
  selectBreeds,
  selectSubBreeds,
} from "@Search/store/selectors";
import { searchActions } from "@Search/store";
import { resultsActions } from "@DisplayResults/store";

export const SearchFormComponent = () => {
  const { t } = useTranslation("search");
  const {
    control,
    handleSubmit,
    setValue,
    formState: { errors, isValid },
  } = useForm<Form>({
    defaultValues: {
      breed: "",
      subBreed: "",
    },
    resolver: yupResolver(SearchValidator),
    mode: "onChange",
  });

  const dispatch = useDispatch();
  const areBreedsAlreadyRequested = useSelector(
    areBreedsAlreadyRequestedSelector
  );
  const breeds = useSelector(selectBreeds);
  const subBreeds = useSelector(selectSubBreeds);

  useEffect(() => {
    if (!areBreedsAlreadyRequested) {
      dispatch(searchActions.getBreeds());
    }
  }, []);

  const onSubmit = (form: Form) => {
    dispatch(searchActions.setSelection(form));
    if (form.subBreed) {
      dispatch(resultsActions.getSubBreedPhotos(form));
    } else {
      dispatch(resultsActions.getBreedPhotos(form.breed));
    }
  };

  return (
    <StyledForm onSubmit={handleSubmit(onSubmit)}>
      <SelectContainer>
        <Controller
          control={control}
          name="breed"
          render={({ field }: any) => (
            <Select
              onSelect={(breed: string) => {
                field.onChange(breed);
                setValue("subBreed", "");
                dispatch(searchActions.getSubBreeds(breed));
              }}
              placeholder={t("searchBreed")}
              options={breeds.map((breed: string) => ({
                name: breed,
                value: breed,
              }))}
              value={field.value}
              onBlur={field.onBlur}
            />
          )}
        />
        <Controller
          control={control}
          name="subBreed"
          render={({ field }: any) => (
            <Select
              onSelect={field.onChange}
              placeholder={t("searchSubBreed")}
              options={subBreeds.map((breed: string) => ({
                name: breed,
                value: breed,
              }))}
              value={field.value}
              onBlur={field.onBlur}
            />
          )}
        />
      </SelectContainer>
      <Button isDisable={!isValid} value={t("search") as string} />
    </StyledForm>
  );
};
