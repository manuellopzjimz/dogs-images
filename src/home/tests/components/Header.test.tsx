import React from "react";
import { Header } from "@Home/components/Header";
import { render } from "@testing-library/react";

describe("HeaderComponent", () => {
  it("should be rendered correctly", () => {
    const { container } = render(
      <Header>
        <div>Left content</div>
        <div>Right content</div>
      </Header>
    );

    expect(container).toMatchSnapshot();
  });
});
