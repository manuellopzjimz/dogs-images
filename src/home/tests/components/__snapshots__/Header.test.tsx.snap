// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent should be rendered correctly 1`] = `
<div>
  <header
    class="sc-bdvvaa bzxsTt"
  >
    <div>
      Left content
    </div>
    <div>
      Right content
    </div>
  </header>
</div>
`;
